lib/std/jsfetch.nim

```diff
@@ -0,0 +1,143 @@
+
+  Response* = ref object  ## https://developer.mozilla.org/en-US/docs/Web/API/Response
+    myBodyUsed, ok, redirected: bool
+    typ {.importjs: "type".}: cstring
+    url, statusText: cstring
+    status: cint
+    headers: Headers
+
+
+proc fetchMethodToCstring(metod: HttpMethod): cstring =
+  ## proc that takes an `HttpMethod` and returns an *Uppercase* `cstring`,
+  ## but *only* for the HTTP Methods that are supported by the fetch API.
+  ## High performance and minimal code compared to just `$(HttpMethod)`.
+  assert metod notin {HttpTrace, HttpOptions, HttpConnect}, "HTTP Method not supported by Fetch API"
+  case metod
+  of HttpHead:   "HEAD".cstring
+  of HttpGet:    "GET".cstring
+  of HttpPost:   "POST".cstring
+  of HttpPut:    "PUT".cstring
+  of HttpDelete: "DELETE".cstring
+  of HttpPatch:  "PATCH".cstring
+  else:          "GET".cstring
+
+func unsafeNewFetchOptions*(metod, body, mode, credentials, cache, referrerPolicy: cstring,
+    keepalive: bool, redirect = "follow".cstring, referrer = "client".cstring, integrity = "".cstring): FetchOptions {.importjs:
+    "{method: #, body: #, mode: #, credentials: #, cache: #, referrerPolicy: #, keepalive: #, redirect: #, referrer: #, integrity: #}".}
+  ## **Unsafe** `newfetchOptions`. Low-level func for optimization.
+
+func newfetchOptions*(metod: HttpMethod, body: cstring,
+    mode: FetchModes, credentials: FetchCredentials, cache: FetchCaches, referrerPolicy: FetchReferrerPolicies,
+    keepalive: bool, redirect = frFollow, referrer = "client".cstring, integrity = "".cstring): FetchOptions =
+  ## Constructor for `FetchOptions`.
+  result = FetchOptions(metod: fetchMethodToCstring(metod), body: body, mode: $mode,
+    credentials: $credentials, cache: $cache, referrerPolicy: $referrerPolicy,
+    keepalive: keepalive, redirect: $redirect , referrer: referrer, integrity: integrity)
+
+func fetchToCstring*(url: cstring): Future[cstring] {.importjs: "fetch(#).then(response => response.text()).then(text => text)".}
+  ## Convenience func for `fetch()` API that returns a `cstring` directly.
+
+func fetchToCstring*(url: cstring, options: FetchOptions): Future[cstring] {.importjs: "fetch(#, #).then(response => response.text()).then(text => text)".}
+  ## Convenience func for `fetch()` API that returns a `cstring` directly.
+
+func fetch*(url: cstring): Future[Response] {.importjs: "fetch(#).then(response => response)".}
+  ## `fetch()` API, simple `GET` only, returns a `Response`.
+
+func fetch*(url: cstring, options: FetchOptions): Future[Response] {.importjs: "(await fetch(#, #).then(response => response))".}
+  ## `fetch()` API that takes a `FetchOptions`, returns a `Response`.
+
+func toCstring*(this: FetchOptions or Response): cstring {.importjs: "JSON.stringify(#)".}
+
+func `$`*(this: FetchOptions or Response): string = $toCstring(this)
+
+
+runnableExamples:
+  import httpcore
+  if defined(nimJsFetchTests):
+    block:
+      let options0: FetchOptions = unsafeNewFetchOptions(
```

Comment:
can you also add an actual fetch example so we know this works?

for example adapting this from https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch
```
fetch('http://example.com/movies.json')
  .then(response => response.json())
  .then(data => console.log(data));
```